api/APIValidation.ts

```diff
@@ -0,0 +1,75 @@
+import { ZodError, ZodTypeAny, z } from "zod";
+import { addLogHandler, consoleLogHandler, logger } from "../logging";
+import { APIMiddleware, TiFAPIInputContext, TiFAPIResponse } from "./TransportTypes";
+
+const log = logger("tif.api.validation")
+addLogHandler(consoleLogHandler())
+
+const zodValidation = async <T>(data: T, schema: ZodTypeAny): Promise<T | "failure"> => {
+  try {
+      return await schema.parseAsync(data);
+  } catch (e) {
+      if (e instanceof ZodError) {
+          log.error("Zod Schema Error Message", {
+              zodError: JSON.parse(e.message)
+          });
+      } else {
+          log.error(e);
+      }
+      return "failure";
+  }
+};
+
+type ValidationResult = "invalid-request" | "unexpected-response" | "invalid-response" | "passed"
+
+type ValidationResultParser = (status: ValidationResult, value: TiFAPIInputContext<any> | TiFAPIResponse<any>) => (TiFAPIResponse<any>)
+
+// NB: Currently needs to be annotated with "any" in order for typescript to mark it compatible with more specific middleware like TiFAPITransport
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export const validateAPICall = (resultParser: ValidationResultParser, validate: "requestOnly" | "responseOnly" | "both" = "both"): APIMiddleware<any> => 
+  async (endpointInput, next) => {
+    const {endpointName, endpointSchema: {input: inputSchema, outputs: outputSchemas, constraints}, body, query, params} = endpointInput
+
+    let result: ValidationResult = "passed"
+
+    let request = {body, query, params}
+
+    if (validate === "requestOnly" || validate === "both") {
+      const validatedRequest = await zodValidation(request, z.object(inputSchema))
+      
+      if (validatedRequest === "failure") {
+        result = "invalid-request"
+        log.error(`Request to TiF API endpoint ${endpointName} is not valid`, request)
```

Comment:
Nit: We could choose to move logging into the `resultParser` function to avoid potential double logging (ie. also logging the error in the `resultParser`).

Reply:
Removed these logs

Reply:
![image](https://github.com/user-attachments/assets/24430136-4c95-4dbf-b292-0b46fb04d3d1)

NVM "endpointName" is also included in these logs so it would be difficult to extract that out.

Reply:
Fair, but maybe the result validation parser should always have access to the `TiFAPIInputContext` in every case since it contains the endpoint name.